### PR TITLE
hls: Initial commit of high-level-synthesis backend

### PR DESCRIPTION
**Pull request by sjalander**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,5 +7,11 @@ check.log
 /bin
 tests/test-runner
 .Makefile.override
+libjlc-hls/include/jlc/
```
Comment (phate) [inline .gitignore:10]: Is this line necessary or is it outdated?

File: .gitignore [hunk 2]
```diff
@@ -7,5 +7,11 @@ check.log
...
 libjlc/include/jlc/llvmpaths.hpp
 docs/html
+libjlc-hls/include/jlc-hls/toolpaths.hpp
+.idea
+cmake-build-debug
```
Comment (phate) [inline .gitignore:15]: Why do we have a cmake-build-debug here? We do not use cmake.

File: jlm-opt/Makefile.sub [hunk 3]
```diff
@@ -13,7 +13,7 @@ jlm-opt-debug: $(JLM_ROOT)/libjive.a $(JLM_ROOT)/libjlm.a $(JLM_ROOT)/bin/jlm-op
 jlm-opt-release: CXXFLAGS += -O3
 jlm-opt-release: $(JLM_ROOT)/libjive.a $(JLM_ROOT)/libjlm.a $(JLM_ROOT)/bin/jlm-opt
 
-$(JLM_ROOT)/bin/jlm-opt: CPPFLAGS += -I$(JLM_ROOT)/libjlm/include -I$(JLM_ROOT)/jlm-opt/include -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir)
+$(JLM_ROOT)/bin/jlm-opt: CPPFLAGS += -I$(JLM_ROOT)/libjlm/include -I$(JLM_ROOT)/jlm-opt/include -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir) $(CPPFLAGS_CIRCT) $(LDFLAGS_CIRCT)
```
Comment (phate) [inline jlm-opt/Makefile.sub:16]: CPPFLASGS_CIRCT and LDFLAGS_CIRCT is nowhere defined.

File: jlm-print/Makefile.sub [hunk 4]
```diff
@@ -12,7 +12,7 @@ jlm-print-debug: $(JLM_ROOT)/libjive.a $(JLM_ROOT)/libjlm.a $(JLM_ROOT)/bin/jlm-
 jlm-print-release: CXXFLAGS += -O3
 jlm-print-release: $(JLM_ROOT)/libjive.a $(JLM_ROOT)/libjlm.a $(JLM_ROOT)/bin/jlm-print
 
-$(JLM_ROOT)/bin/jlm-print: CPPFLAGS += -I$(JLM_ROOT)/libjlm/include -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir)
+$(JLM_ROOT)/bin/jlm-print: CPPFLAGS += -I$(JLM_ROOT)/libjlm/include -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir) $(CPPFLAGS_CIRCT) $(LDFLAGS_CIRCT)
```
Comment (phate) [inline jlm-print/Makefile.sub:15]: CPPFLASGS_CIRCT and LDFLAGS_CIRCT is nowhere defined.

File: libjlc/Makefile.sub [hunk 5]
```diff
@@ -37,7 +37,7 @@ libjlc-debug: $(JLM_ROOT)/libjlc.a
 libjlc-release: CXXFLAGS += -O3
 libjlc-release: $(JLM_ROOT)/libjlc.a
 
-$(JLM_ROOT)/libjlc.a: CPPFLAGS += -I$(JLM_ROOT)/libjlc/include -I$(JLM_ROOT)/libjlm/include -I$(shell $(LLVMCONFIG) --includedir)
+$(JLM_ROOT)/libjlc.a: CPPFLAGS += -I$(JLM_ROOT)/libjlc/include -I$(JLM_ROOT)/libjlm/include -I$(shell $(LLVMCONFIG) --includedir) $(CPPFLAGS_CIRCT) $(LDFLAGS_CIRCT)
```
Comment (phate) [inline libjlc/Makefile.sub:40]: CPPFLASGS_CIRCT and LDFLAGS_CIRCT is nowhere defined.

File: libjlm/Makefile.sub [hunk 6]
```diff
@@ -72,7 +102,7 @@ libjlm-debug: $(JLM_ROOT)/libjlm.a
 libjlm-release: CXXFLAGS += -O3
 libjlm-release: $(JLM_ROOT)/libjlm.a
 
-$(JLM_ROOT)/libjlm.a: CPPFLAGS += -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir) -I$(JLM_ROOT)/libjlm/include
+$(JLM_ROOT)/libjlm.a: CPPFLAGS += -I$(JLM_ROOT)/libjive/include -I$(shell $(LLVMCONFIG) --includedir) -I$(JLM_ROOT)/libjlm/include $(CPPFLAGS_CIRCT) $(LDFLAGS_CIRCT)
```
Comment (phate) [inline libjlm/Makefile.sub:105]: CPPFLASGS_CIRCT and LDFLAGS_CIRCT is nowhere defined.

File: libjlm/include/jlm/ir/hls/hls.h [hunk 7]
```diff
@@ -0,0 +1,560 @@
+//
+// Created by david on 7/2/21.
+//
+
+#ifndef JLM_IR_HLS_HLS_HPP
+#define JLM_IR_HLS_HLS_HPP
```
Comment (phate) [inline libjlm/include/jlm/ir/hls/hls.h:6]: The file should be called hls.hpp and not hls.h

File: libjlm/include/jlm/backend/hls/rvsdg2rhls/inline.hpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
+//
+// Created by david on 7/2/21.
+//
+
+#ifndef JLM_BACKEND_HLS_RVSDG2RHLS_INLINE_HPP
+#define JLM_BACKEND_HLS_RVSDG2RHLS_INLINE_HPP
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rvsdg2rhls/inline.hpp:6]: This file can be completely removed or not? Nothing really here.

File: libjlm/src/backend/hls/rvsdg2rhls/inline.cpp [hunk 9]
```diff
@@ -0,0 +1,5 @@
+//
```
Comment (phate) [inline libjlm/src/backend/hls/rvsdg2rhls/inline.cpp:1]: Same here. Please remove.

File: libjlc-hls/include/jlc-hls/cmdline.hpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2018 Nico Reißmann <nico.reissmann@gmail.com>
```
Comment (phate) [inline libjlc-hls/include/jlc-hls/cmdline.hpp:2]: I did not write this file. Could be it that this is just a blunt copy of jlc's cmdline.hpp?

File: libjlc-hls/include/jlc-hls/cmdline.hpp [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_JLC_CMDLINE_HPP
```
Comment (phate) [inline libjlc-hls/include/jlc-hls/cmdline.hpp:6]: Macro does not fit the file path.

File: libjlc-hls/src/cmdline.cpp [hunk 12]
```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2018 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
```
Comment (phate) [inline libjlc-hls/src/cmdline.cpp:5]: Missing header guard.

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen-circt.hpp [hunk 13]
```diff
@@ -0,0 +1,218 @@
+//
+// Created by Magnus Sjalander 26 December 2021
+//
+
+#ifndef JLM_BACKEND_HLS_RHLS2FIRRTL_MLIRGEN_HPP
+#define JLM_BACKEND_HLS_RHLS2FIRRTL_MLIRGEN_HPP
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen-circt.hpp:6]: Guard does not fit file name.

File: libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen-llvm.hpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
+//
+// Created by Magnus Sjalander 26 December 2021
+//
+
+#ifndef JLM_BACKEND_HLS_RHLS2FIRRTL_MLIRGEN_HPP
+#define JLM_BACKEND_HLS_RHLS2FIRRTL_MLIRGEN_HPP
```
Comment (phate) [inline libjlm/include/jlm/backend/hls/rhls2firrtl/mlirgen-llvm.hpp:6]: Guard does not fit file name.

**Review by phate — COMMENTED** [review]
Could you also add the dependencies for the entire HLS stuff to the README.md

File: libjlc-hls/src/cmdline.cpp (hunk 12 above)
Comment (sjalander) [reply libjlc-hls/src/cmdline.cpp:5]: Did you comment on the wrong file?

File: libjlc-hls/include/jlc-hls/cmdline.hpp (hunk 10 above)
Comment (sjalander) [reply libjlc-hls/include/jlc-hls/cmdline.hpp:2]: Yes, so it is mostly you that has written it

**Comment by sjalander** [comment]
We do not want to advertise the HLS support yet

**Pull request closed without merging**